### PR TITLE
Beest

### PR DESCRIPTION
**Pull request by TizianoPerrucci**
### Step 3: Did you add / update tests for your changes in the right area?
- [ ] Unit/Component test
- [ ] Integration test
- [ ] E2E/Molecule test

### Step 4: I'm confident that everything is properly tested:
I got a PO / QA Approval by:
- [ ] Name

### Step 5: Did you add release notes describing the changes you made?
- [ ] Yes

### Step 6: Can we ship this feature to production?
- [ ] Yes, I'm proud of my work. Ship it! :ship:

### CODE REVIEW

File: beest/cmd/driver/pytest.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	return cmd
+}
+
+func runPyTestCmd(ctx context.Context, cmd *exec.Cmd) error {
+	var errBuf strings.Builder
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = io.MultiWriter(os.Stderr, &errBuf)
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{
```
Comment (zandrewitte) [inline beest/cmd/driver/pytest.go:85]: Interestingly Windows has a totally different model for `syscall.SysProcAttr`. I doubt anyone will run it on windows I just noticed it on my machine now, but it's probably good to add a `!windows` build flag.

Comment (zandrewitte) [reply beest/cmd/driver/pytest.go:85]: I guess that also means you have to implement a `pytest_windows.go` but just throwing Unimplemented error for now should be fine

File: beest/cmd/prepare.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"beest/cmd/driver"
+	"beest/cmd/step"
+	"github.com/spf13/cobra"
+)
+
+var prepareCmd = &cobra.Command{
+	Use:   "prepare [scenario]",
+	Short: "Deploy the integrations configured as part of the yard",
```
Comment (zandrewitte) [inline beest/cmd/prepare.go:11]: Let's use the same bee terminology here to make it consistent. Can't remember if this is done for Bees or Hives? 😄

File: beest/keeper.dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
+FROM stackstate/stackstate-agent-2-test:beest-keeper-base-20220108
```
Comment (zandrewitte) [inline beest/keeper.dockerfile:1]: Same thing about the date here. We should parameterize it

**Comment by zandrewitte** [comment]
Most of the exported functions are missing comments, so I expect if you run a linter you will get quite a bit of complaining.

File: .gitlab-ci-agent.yml [hunk 4]
```diff
@@ -672,6 +672,63 @@ pre_release_cluster_agent_image:
...
+# Core Structure for all the beest steps
+.beest: &beest
+  image: stackstate/stackstate-agent-2-test:beest-keeper-base-20220108
+  interruptible: true
+  dependencies: []
+  before_script:
+    - mkdir /go/src/app
+    - cp -R $CI_PROJECT_DIR/beest/* /go/src/app
+    - cd /go/src/app
+    - go build .
```
[the remark this answers is not in the document]
Comment (TizianoPerrucci) [reply .gitlab-ci-agent.yml:685]: lets talk about this

File: beest/Makefile [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+GIT_BRANCH = $(shell git rev-parse --abbrev-ref HEAD)
+BRANCH = $(shell echo ${GIT_BRANCH} | cut -d'/' -f 2- | tr [:upper:] [:lower:])
+
+STACKSTATE_BRANCH            ?= 4.5.1
+MAJOR_VERSION                ?= 2
+STS_AWS_TEST_BUCKET          ?= stackstate-agent-2-test
+STS_DOCKER_TEST_REPO         ?= stackstate-agent-2-test
+STS_DOCKER_TEST_REPO_CLUSTER ?= stackstate-cluster-agent-test
+
+BASE_KEEPER_IMAGE  = stackstate/stackstate-agent-2-test:beest-keeper-base-20220108
```
[the remark this answers is not in the document]
Comment (TizianoPerrucci) [reply beest/Makefile:26]: this is the base image, will change only if we add a new tool. so generally it does not change that often

File: beest/bootstrap.sh [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+source ~/.bashrc
+eval "$(direnv hook bash)"
+direnv allow
+
+go install github.com/spf13/cobra/cobra
```
[the remark this answers is not in the document]
Comment (TizianoPerrucci) [reply beest/bootstrap.sh:7]: cobra is definitely versioned in go.mod `github.com/spf13/cobra v1.2.1`, this is just providing their cli (https://github.com/spf13/cobra/blob/master/cobra/README.md) which is simply a generator which allows to create scaffolding quickly for a new command

File: beest/cmd/cleanup.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+import (
+	"beest/cmd/driver"
+	"beest/cmd/step"
+	"github.com/spf13/cobra"
+)
+
+var cleanupCmd = &cobra.Command{
+	Use:   "cleanup [scenario]",
+	Short: "Uninstall all the configured integrations",
```
[the remark this answers is not in the document]
Comment (TizianoPerrucci) [reply beest/cmd/cleanup.go:11]: agreed

File: beest/cmd/create.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+var createCmd = &cobra.Command{
+	Use:   "create [scenario]",
+	Short: "Provision the yard used by a certain scenario",
+	Args:  cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		scenario = choseScenario(args[0])
+
+		create := step.Create(scenario.Yard.path())
+		doCreate(create, !assumeYes)
```
[the remark this answers is not in the document]
Comment (TizianoPerrucci) [reply beest/cmd/create.go:17]: correct, by default you just run a `create` will prompt for confirmation, while the `test` command should not prompt

File: beest/cmd/destroy.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+var destroyCmd = &cobra.Command{
+	Use:   "destroy [scenario]",
+	Short: "Destroy all resources associated with the yard",
+	Args:  cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		scenario = choseScenario(args[0])
+
+		create := step.Create(scenario.Yard.path())
+		destroy := step.Destroy(create)
```
[the remark this answers is not in the document]
Comment (TizianoPerrucci) [reply beest/cmd/destroy.go:17]: we can talk about it

File: beest/cmd/driver/pytest.go (hunk 1 above)
Comment (TizianoPerrucci) [reply beest/cmd/driver/pytest.go:85]: agreed

File: beest/cmd/prepare.go (hunk 2 above)
Comment (TizianoPerrucci) [reply beest/cmd/prepare.go:11]: agreed

File: beest/keeper.dockerfile (hunk 3 above)
Comment (TizianoPerrucci) [reply beest/keeper.dockerfile:1]: agreed

File: beest/cmd/driver/terraform.go [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+
+func newTerraform(ctx TerraformContext) *TerraformState {
+	log.Println(fmt.Sprintf("Preparing Terraform to run from: %s", ctx.WorkingDir()))
+
+	tf, err := tfexec.NewTerraform(ctx.WorkingDir(), DefaultTerraformBinary)
+	if err != nil {
+		log.Fatalf("Error running NewTerraform: %s", err)
+	}
+
+	// TODO debug
```
Comment (julianosk) [inline beest/cmd/driver/terraform.go:57]: expand this TODO so we know what is your idea

File: beest/cmd/root.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+	if !ok {
+		log.Fatalf("Mandatory environment variable missing: %s", AgentCurrentBranchEnvVar)
+	}
+}
+
+func init() {
+	cobra.OnInitialize(initConfig)
+
+	// persistent flags global for the application
+	rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.beest.yaml)")
```
Comment (julianosk) [inline beest/cmd/root.go:57]: Not needed for now. drop it

File: beest/cmd/root.go [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+	rootCmd.PersistentFlags().StringVar(&yardId, YardIdFlag, "", "used as prefix for resource names")
+}
+
+// initConfig reads in config file and ENV variables if set.
+func initConfig() {
```
Comment (julianosk) [inline beest/cmd/root.go:63]: Also not needed.

File: beest/keeper.base.dockerfile [hunk 13]
```diff
@@ -0,0 +1,69 @@
+FROM golang:1.17.3-bullseye
```
Comment (julianosk) [inline beest/keeper.base.dockerfile:1]: Move this file to `.cd-builders`

File: beest/sut/bees/tasks/install-and-update/install-stackstate-agent.yml [hunk 14]
```diff
@@ -0,0 +1,13 @@
+---
```
Comment (julianosk) [inline beest/sut/bees/tasks/install-and-update/install-stackstate-agent.yml:1]: Why do we need these ansible files? Remove

File: beest/sut/bees/tasks/install-and-update/updates-ubuntu18-automatic.yml [hunk 15]
```diff
@@ -0,0 +1,8 @@
+---
```
Comment (julianosk) [inline beest/sut/bees/tasks/install-and-update/updates-ubuntu18-automatic.yml:1]: Also remove

File: beest/sut/yards/eks-receiver/providers.tf [hunk 16]
```diff
@@ -0,0 +1,31 @@
+terraform {
+  backend "s3" {
+    bucket = "lupulus-sandbox-terraform-state"
+    key    = "tiziano-aws-eks-1-21-shard.terraform.tfstate"
```
Comment (julianosk) [inline beest/sut/yards/eks-receiver/providers.tf:4]: Maybe use a more generic name for now

**Review by julianosk — CHANGES_REQUESTED** [review]
A few remarks.

File: .gitlab-ci-agent.yml [hunk 17]
```diff
@@ -672,6 +672,63 @@ pre_release_cluster_agent_image:
...
+# Primary create steps for all the beest builds
+.beest_create: &beest_create
+  <<: *beest_artifacts
+  stage: prepare
+  script:
+    - ls -la /root/.kube/* || true
+    - cd /go/src/app
+    - beest create $BEEST_SCENARIO -y
+    - cat /root/.kube/*
+    - ansible -c local -i localhost, -m setup localhost
```
[the remark this answers is not in the document]
Comment (TizianoPerrucci) [reply .gitlab-ci-agent.yml]: removed them, was for debugging

**Comment by TizianoPerrucci** [comment]
> Most of the exported functions are missing comments, so I expect if you run a linter you will get quite a bit of complaining.

create a ticket for it https://stackstate.atlassian.net/browse/STAC-15550

File: .gitlab-ci-agent.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (TizianoPerrucci) [reply .gitlab-ci-agent.yml:685]: created a ticket for this https://stackstate.atlassian.net/browse/STAC-15552

I'll schedule also a meeting with us and Martin

File: beest/cmd/driver/terraform.go (hunk 10 above)
Comment (TizianoPerrucci) [reply beest/cmd/driver/terraform.go:57]: I'll remove and created a ticket to introduce a CLI flag to output debug information https://stackstate.atlassian.net/browse/STAC-15551

File: beest/cmd/root.go (hunk 11 above)
Comment (TizianoPerrucci) [reply beest/cmd/root.go:57]: agreed

File: beest/sut/bees/tasks/install-and-update/install-stackstate-agent.yml (hunk 14 above)
Comment (TizianoPerrucci) [reply beest/sut/bees/tasks/install-and-update/install-stackstate-agent.yml:1]: agreed

File: beest/sut/bees/tasks/install-and-update/updates-ubuntu18-automatic.yml (hunk 15 above)
Comment (TizianoPerrucci) [reply beest/sut/bees/tasks/install-and-update/updates-ubuntu18-automatic.yml:1]: agreed

File: beest/sut/yards/eks-receiver/providers.tf (hunk 16 above)
Comment (TizianoPerrucci) [reply beest/sut/yards/eks-receiver/providers.tf:4]: agreed
also created this ticket https://stackstate.atlassian.net/browse/STAC-15553

**Comment by TizianoPerrucci** [comment]
open from the gitlab side

**Pull request closed without merging**